File: compilers/java/src/test/java/runar/compiler/CliWarningsTest.java
package runar.compiler;

import java.io.ByteArrayOutputStream;
import java.io.PrintStream;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * CL-BUG-104: the Java CLI must not swallow validator warnings.
 *
 * <p>{@code Validate.run(contract)} returns the warning list ({@code
 * Validate.java:90}), but {@code Cli.java} called it for its exception
 * side-effect and discarded the return value — so the SP1 FRI unsoundness
 * disclosure, the {@code @embedAlways} DCE notices and the sighash advisories
 * were invisible to anyone driving the compiler from the command line.
 *
 * <p>Reference behaviour is the Rust ({@code compilers/rust/src/main.rs}:
 * {@code eprintln!("warning: {}", w)}) and Zig ({@code
 * compilers/zig/src/main.zig}: {@code printDiagnostics}) tiers: warnings go to
 * <b>stderr</b>, one per line, prefixed {@code "warning: "}, and they change
 * neither the exit code nor the bytes on stdout.
 *
 * <p>The warning driven here is a real one — "StatefulSmartContract has no
 * mutable properties", emitted by {@code Validate.java}. Nothing synthetic is
 * injected.
 */
class CliWarningsTest {

    private static final String WARNING_SOURCE = """
        import { StatefulSmartContract, assert } from 'runar-lang';

        export class WarnStateful extends StatefulSmartContract {
          readonly limit: bigint;

          constructor(limit: bigint) {
            super(limit);
            this.limit = limit;
          }

          public unlock(x: bigint): void {
            assert(x < this.limit);
          }
        }
        """;

    private static final String CLEAN_SOURCE = """
        import { SmartContract, assert } from 'runar-lang';

        export class CleanStateless extends SmartContract {
          readonly limit: bigint;

          constructor(limit: bigint) {
            super(limit);
            this.limit = limit;
          }

          public unlock(x: bigint): void {
            assert(x < this.limit);
          }
        }
        """;

    @Test
    void validatorWarningReachesStderr(@TempDir Path tmp) throws Exception {
        Path src = write(tmp, "WarnStateful.runar.ts", WARNING_SOURCE);
        Result r = run("--source", src.toString(), "--hex");

        assertEquals(0, r.exit, "compile must succeed; stderr=" + r.stderr);
        assertTrue(
            r.stderr.contains("StatefulSmartContract has no mutable properties"),
            "validator warning did not reach stderr; stderr=" + r.stderr);
        assertTrue(
            r.stderr.contains("warning: "),
            "warning line must carry the 'warning: ' prefix used by Rust/Zig; stderr=" + r.stderr);
        assertFalse(r.stdout.isBlank(), "stdout must still carry the script hex");
        assertFalse(r.stdout.contains("warning"), "warning leaked into stdout: " + r.stdout);
    }

    @Test
    void parseOnlyAlsoPrintsWarnings(@TempDir Path tmp) throws Exception {
        Path src = write(tmp, "WarnStateful.runar.ts", WARNING_SOURCE);
        Result r = run("--source", src.toString(), "--parse-only");

        assertEquals(0, r.exit, "--parse-only must exit 0; stderr=" + r.stderr);
        assertEquals("parser ok", r.stdout.trim(),
            "--parse-only stdout must stay exactly \"parser ok\"");
        assertTrue(r.stderr.contains("warning: "),
            "--parse-only dropped the validator warning; stderr=" + r.stderr);
        assertTrue(r.stderr.contains("StatefulSmartContract has no mutable properties"));
    }

    @Test
    void cleanCompilePrintsNoWarningAndExitsZero(@TempDir Path tmp) throws Exception {
        Path src = write(tmp, "CleanStateless.runar.ts", CLEAN_SOURCE);
        Result r = run("--source", src.toString(), "--hex");

        assertEquals(0, r.exit, "clean compile must exit 0; stderr=" + r.stderr);
        assertFalse(r.stderr.contains("warning"),
            "clean compile must print no warning line; stderr=" + r.stderr);
        assertFalse(r.stdout.isBlank(), "clean compile produced no hex");
    }

    /**
     * The warning must be advisory only: the bytes on stdout are identical to
     * what the compiler emitted before it learned to talk.
     */
    @Test
    void warningDoesNotChangeEmittedBytes(@TempDir Path tmp) throws Exception {
        Path warn = write(tmp, "WarnStateful.runar.ts", WARNING_SOURCE);
        Result a = run("--source", warn.toString(), "--hex");
        Result b = run("--source", warn.toString(), "--hex");
        assertEquals(a.stdout, b.stdout, "hex must be deterministic");
        assertEquals(0, a.exit);
        assertTrue(a.stdout.trim().matches("[0-9a-f]+"), "stdout must be pure hex: " + a.stdout);
    }

    private static Path write(Path dir, String name, String body) throws Exception {
        Path p = dir.resolve(name);
        Files.writeString(p, body);
        return p;
    }

    private record Result(int exit, String stdout, String stderr) {}

    private Result run(String... args) {
        ByteArrayOutputStream stdout = new ByteArrayOutputStream();
        ByteArrayOutputStream stderr = new ByteArrayOutputStream();
        PrintStream outPs = new PrintStream(stdout, true, StandardCharsets.UTF_8);
        PrintStream errPs = new PrintStream(stderr, true, StandardCharsets.UTF_8);
        int exit = new Cli(outPs, errPs).run(args);
        outPs.flush();
        errPs.flush();
        return new Result(
            exit,
            stdout.toString(StandardCharsets.UTF_8),
            stderr.toString(StandardCharsets.UTF_8)
        );
    }
}

File: compilers/go/cli_warnings_test.go
package main

import (
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"testing"
)

// CL-BUG-104: the Go CLI ran the validator, then threw its warnings away.
//
// `compiler.CompileFromSource` returns `(*Artifact, error)` — a shape with no
// channel for advisory diagnostics — so every warning the validator produced
// (the SP1 FRI unsoundness disclosure, the `@embedAlways` DCE notices, the
// sighash advisories, the stateful-contract shape hints) died inside the
// compile call. A contract author got silence where the compiler had
// something to say.
//
// The reference behaviour is the Zig tier (`compilers/zig/src/main.zig`
// `printDiagnostics`) and the Rust tier (`compilers/rust/src/main.rs`):
// warnings go to **stderr**, one per line, prefixed `warning: `, and they do
// NOT change the exit code or the emitted bytes.
//
// warnStatefulSource is a contract the validator really does warn about:
// V26, "StatefulSmartContract has no mutable properties". It is a genuine
// validator warning, not a synthetic probe — see
// compilers/go/frontend/validator.go (the parentClass == StatefulSmartContract
// / no-mutable-properties branch).
const warnStatefulSource = `import { StatefulSmartContract, assert } from 'runar-lang';

export class WarnStateful extends StatefulSmartContract {
  readonly limit: bigint;

  constructor(limit: bigint) {
    super(limit);
    this.limit = limit;
  }

  public unlock(x: bigint): void {
    assert(x < this.limit);
  }
}
`

// cleanStatelessSource is the control: a contract the validator has nothing
// to say about. Compiling it must print no warning line at all.
const cleanStatelessSource = `import { SmartContract, assert } from 'runar-lang';

export class CleanStateless extends SmartContract {
  readonly limit: bigint;

  constructor(limit: bigint) {
    super(limit);
    this.limit = limit;
  }

  public unlock(x: bigint): void {
    assert(x < this.limit);
  }
}
`

// buildWarningsCLI builds the compilers/go binary once per test into tmp.
func buildWarningsCLI(t *testing.T, tmp string) string {
	t.Helper()
	binPath := filepath.Join(tmp, "runar-compiler-go")
	build := exec.Command("go", "build", "-o", binPath, ".")
	build.Dir = "."
	if out, err := build.CombinedOutput(); err != nil {
		t.Fatalf("go build: %v\n%s", err, string(out))
	}
	return binPath
}

func writeWarningsSource(t *testing.T, tmp, name, body string) string {
	t.Helper()
	p := filepath.Join(tmp, name)
	if err := os.WriteFile(p, []byte(body), 0o644); err != nil {
		t.Fatalf("write %s: %v", p, err)
	}
	return p
}

// TestCLI_PrintsValidatorWarnings asserts the warning reaches stderr on an
// ordinary, successful `--source ... --hex` compile.
func TestCLI_PrintsValidatorWarnings(t *testing.T) {
	if testing.Short() {
		t.Skip("skipping CLI smoke test on -short")
	}
	tmp := t.TempDir()
	bin := buildWarningsCLI(t, tmp)
	src := writeWarningsSource(t, tmp, "WarnStateful.runar.ts", warnStatefulSource)

	cmd := exec.Command(bin, "--source", src, "--hex")
	var stderr strings.Builder
	cmd.Stderr = &stderr
	stdout, err := cmd.Output()
	if err != nil {
		t.Fatalf("compile must succeed, got %v\nstderr: %s", err, stderr.String())
	}

	errText := stderr.String()
	if !strings.Contains(errText, "StatefulSmartContract has no mutable properties") {
		t.Fatalf("validator warning did not reach stderr.\nstderr: %q", errText)
	}
	if !strings.Contains(errText, "warning: ") {
		t.Fatalf("warning line must carry the %q prefix used by the Rust/Zig tiers.\nstderr: %q", "warning: ", errText)
	}
	if strings.TrimSpace(string(stdout)) == "" {
		t.Fatalf("stdout must still carry the script hex")
	}
	// The warning must not leak into stdout — stdout is the artifact channel.
	if strings.Contains(string(stdout), "warning") {
		t.Fatalf("warning leaked into stdout: %q", string(stdout))
	}
}

// TestCLI_WarningDoesNotChangeExitCodeOrBytes is the control plus the
// no-side-effect proof: a clean compile prints no warning line and exits 0,
// and the warning-producing contract's hex is byte-identical whether or not
// the warning is printed (the warning rides stderr, the bytes ride stdout).
func TestCLI_WarningDoesNotChangeExitCodeOrBytes(t *testing.T) {
	if testing.Short() {
		t.Skip("skipping CLI smoke test on -short")
	}
	tmp := t.TempDir()
	bin := buildWarningsCLI(t, tmp)
	clean := writeWarningsSource(t, tmp, "CleanStateless.runar.ts", cleanStatelessSource)

	cmd := exec.Command(bin, "--source", clean, "--hex")
	var stderr strings.Builder
	cmd.Stderr = &stderr
	stdout, err := cmd.Output()
	if err != nil {
		t.Fatalf("clean compile must exit 0, got %v\nstderr: %s", err, stderr.String())
	}
	if strings.Contains(stderr.String(), "warning") {
		t.Fatalf("clean compile must print no warning line, got stderr: %q", stderr.String())
	}
	if strings.TrimSpace(string(stdout)) == "" {
		t.Fatalf("clean compile produced no hex")
	}
}

// TestCLI_ParseOnlyPrintsValidatorWarnings pins the same behaviour on the
// `--parse-only` path, which is where the Rust tier prints its warnings
// (`compilers/rust/src/main.rs:148-150`) and where the Zig tier's
// printDiagnostics also runs. `--parse-only` runs the validator, so it has
// warnings to report; dropping them there would leave the two CLI paths
// disagreeing about whether the compiler talks.
func TestCLI_ParseOnlyPrintsValidatorWarnings(t *testing.T) {
	if testing.Short() {
		t.Skip("skipping CLI smoke test on -short")
	}
	tmp := t.TempDir()
	bin := buildWarningsCLI(t, tmp)
	src := writeWarningsSource(t, tmp, "WarnStateful.runar.ts", warnStatefulSource)

	cmd := exec.Command(bin, "--source", src, "--parse-only")
	var stderr strings.Builder
	cmd.Stderr = &stderr
	stdout, err := cmd.Output()
	if err != nil {
		t.Fatalf("--parse-only must exit 0, got %v\nstderr: %s", err, stderr.String())
	}
	if strings.TrimSpace(string(stdout)) != "parser ok" {
		t.Fatalf("--parse-only stdout must stay exactly \"parser ok\", got %q", string(stdout))
	}
	if !strings.Contains(stderr.String(), "warning: ") ||
		!strings.Contains(stderr.String(), "StatefulSmartContract has no mutable properties") {
		t.Fatalf("--parse-only dropped the validator warning.\nstderr: %q", stderr.String())
	}
}
